Add library that provides a weighted tree-based fairness

dongahn opened the pull request:
This PR add libweighted_tree that implements
a weighted tree-based fairness algorithm.

Provide fairness calculation based on the high-level
"hierarchical scheduling" concept explained in a paper
like https://people.eecs.berkeley.edu/~alig/papers/h-drf.pdf.
The algorithm is also similar to SLURM's fair tree
as explained in https://slurm.schedmd.com/fair_tree.html.

This PR is the first that introduces compiled code
into flux-accounting, as such we add Autotool support
for config-time checks etc.
We will ultimately transition to Autotool style similar
to other Flux projects like flux-sched.

The library code is at `src/fairness/weighted_tree`
and the unit test is at `src/fairness/weighted_tree/test`.

For these unit tests, this PR imported `libtap`
into `src/common/libtap`.

dongahn commented:
This PR is ready for your review. I don't have much time for awhile so it would be great if you can review this PR as is and let me know what are the minimal amount of things that need to be done to land it before our next release target (end of this week.) If you want to try this w/ a docker container:

```console
% docker run -ti -v ${YOUR_LOCAL_FLUX_ACCOUNTING_DIR}:/usr/src/: fluxrm/flux-core
```

Inside the docker container

```console
ƒ(s=1,d=0) fluxuser@951dad59a9f2:/usr/src$ ./autogen.sh
Running libtoolize --automake --copy ...
Running autoreconf --verbose --install
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal -I config
autoreconf: configure.ac: tracing
autoreconf: running: libtoolize --copy
autoreconf: running: /usr/bin/autoconf
autoreconf: running: /usr/bin/autoheader
autoreconf: running: automake --add-missing --copy --no-force
autoreconf: Leaving directory `.'
Now run ./configure.

ƒ(s=1,d=0) fluxuser@951dad59a9f2:/usr/src$ ./configure
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes

<CUT>

ƒ(s=1,d=0) fluxuser@951dad59a9f2:/usr/src$ make
make  all-recursive
make[1]: Entering directory '/usr/src'
Making all in src
make[2]: Entering directory '/usr/src/src'
Making all in common
make[3]: Entering directory '/usr/src/src/common'
Making all in libtap
make[4]: Entering directory '/usr/src/src/common/libtap'
make[4]: Nothing to be done for 'all'.
make[4]: Leaving directory '/usr/src/src/common/libtap'
make[4]: Entering directory '/usr/src/src/common'
make[4]: Nothing to be done for 'all-am'.
make[4]: Leaving directory '/usr/src/src/common'
make[3]: Leaving directory '/usr/src/src/common'
Making all in fairness
make[3]: Entering directory '/usr/src/src/fairness'
Making all in weighted_tree
make[4]: Entering directory '/usr/src/src/fairness/weighted_tree'
Making all in .
make[5]: Entering directory '/usr/src/src/fairness/weighted_tree'
  CXX      ../account/libweighted_tree_la-account.lo
  CXX      libweighted_tree_la-weighted_tree.lo
  CXX      libweighted_tree_la-weighted_walk.lo
  CXXLD    libweighted_tree.la

<CUT>

ƒ(s=1,d=0) fluxuser@951dad59a9f2:/usr/src$ make check
Making check in src
make[1]: Entering directory '/usr/src/src'
Making check in common
make[2]: Entering directory '/usr/src/src/common'
Making check in libtap
make[3]: Entering directory '/usr/src/src/common/libtap'
make  libtap.la
make[4]: Entering directory '/usr/src/src/common/libtap'
  CC       tap.lo
  CCLD     libtap.la
ar: `u' modifier ignored since `D' is the default (see `U')
make[4]: Leaving directory '/usr/src/src/common/libtap'
make[3]: Leaving directory '/usr/src/src/common/libtap'
make[3]: Entering directory '/usr/src/src/common'
make[3]: Nothing to be done for 'check-am'.
make[3]: Leaving directory '/usr/src/src/common'
make[2]: Leaving directory '/usr/src/src/common'
Making check in fairness
make[2]: Entering directory '/usr/src/src/fairness'
Making check in weighted_tree
make[3]: Entering directory '/usr/src/src/fairness/weighted_tree'
Making check in .
make[4]: Entering directory '/usr/src/src/fairness/weighted_tree'
make[4]: Nothing to be done for 'check-am'.
make[4]: Leaving directory '/usr/src/src/fairness/weighted_tree'
Making check in test
make[4]: Entering directory '/usr/src/src/fairness/weighted_tree/test'
make  weighted_tree_test01
make[5]: Entering directory '/usr/src/src/fairness/weighted_tree/test'
  CXX      weighted_tree_test01-weighted_tree_test01.o
  CXX      weighted_tree_test01-weighted_tree_load.o
  CXXLD    weighted_tree_test01
make[5]: Leaving directory '/usr/src/src/fairness/weighted_tree/test'
make  check-TESTS
make[5]: Entering directory '/usr/src/src/fairness/weighted_tree/test'
make[6]: Entering directory '/usr/src/src/fairness/weighted_tree/test'
PASS: weighted_tree_test01
============================================================================
Testsuite summary for flux-accounting 11e41f2
============================================================================
# TOTAL: 1
# PASS:  1
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================

```

cmoussa1 reviewed (commented):
Thanks @dongahn! I just have one general question about how both of our PR's will fit together:

In the immediate future, will the INPUT for your code require a `.csv` file of user accounts and their shares, similar to what you have in `src/fairness/weighted_tree/test/accounts_data/`?

```
0,root,%^+_nouser,1000,133
1,account1,%^+_nouser,1000,120
2,account1,leaf.1.1,10000,100
2,account1,leaf.1.2,1000,10
2,account1,leaf.1.3,100000,10
1,account2,%^+_nouser,100,12
2,account2,leaf.2.1,0,10
2,account2,leaf.2.2,0,1
```

I ask because my PR currently generates a multi-line string, with indents (" ") indicating the level within the hierarchy. I don't think generating a `.csv` will be too different, though, so I wouldn't mind converting this INPUT to reflect what you have in that directory.

dongahn commented:
@cmoussa1: this is just for test code and the actual weighted tree implementation will not depend on a CSV input at all. To add unit test for integration between your code and mine should be done in a way to make your work easy.

dongahn commented:
FWIW, I used the following GAWK script to parse the out from `sshare -la` into the target CSV format and found some interesting nuance (as being discussed in our mailing list).

```awk
BEGIN  { OFS=","; FIELDWIDTHS = "21 11 11 12 12 14 10"}
NR > 2 {
    sum = sum + 1
    account = $1
    user = $2
    shares = $3
    usage = $5
    fshare = $7
    level = 0
    if (match(account, /^ +/)) level = RLENGTH
    sub(/^ +/, "", account)
    sub(/ +$/, "", account)
    sub(/^ +/, "", user)
    sub(/ +$/, "", user)
    sub(/^ +/, "", shares)
    sub(/ +$/, "", shares)
    sub(/^ +/, "", usage)
    sub(/ +$/, "", usage)
    sub(/^ +/, "", fshare)
    sub(/ +$/, "", fshare)
    if (user == "")
        user = "%^+_nouser"
    #if (user != "%^+_nouser")
    #    user = "user" sum
    if (shares == "")
        shares = 0
    if (usage == "")
        usage = 0
    if (fshare == "")
        fshare = 0
    print level,account,user,shares,usage,fshare
}
```

cmoussa1 reviewed (approved):
LGTM. Thanks @dongahn!

ryanday36 commented:
@dongahn asked me to take a look at this too. I mostly looked at the tests to see if we had the same understanding of the tiebreaker logic and such, and they looked good to me.

dongahn commented:
@ryanday36: Great. Thank you. 

@SteVwonder: if you can give a quick look at it, this can go in for the next release.

SteVwonder reviewed (approved):
LGTM! Thanks @dongahn for putting this together!

Just one minor request for a citation/reference in the comments of the FP equality check:

dongahn commented on src/fairness/weighted_tree/weighted_tree.cpp:
```diff
@@ -0,0 +1,269 @@
+
+bool weighted_tree_node_t::is_equal (double a, double b) const
+{
+    double threshold = std::numeric_limits<double>::epsilon ()
```
Yes. I will do that. 

My Utah collaborators sent me: pg 443 of the collision detection book: http://realtimecollisiondetection.net/books/rtcd/.

Thank you for the review @SteVwonder!

dongahn commented:
OK. I added the suggested comment. I also found some trailing whitespaces which I fixed as well. Added the MWP label as three of you reviewed and approved it.
